## Gather

In [3]:
import pandas as pd
import numpy as np
import requests 
import io
import os
import json
import re 
from timeit import default_timer as timer
from datetime import datetime

In [2]:
import tweepy

consumer_key = '1lFwLDsXPWwt3jqBgEzIgXVuP'
consumer_secret = 'imE7foseU4y8bWpCbBhLXUhcrjzdiOPC0C2tGt9mU0o4m8Xk08'
access_token = '103315165-UIuRdrvqcw5zws0PypTGrlod6Dj2eVnvUHFUGSAX'
access_secret = 'c55H6gf4YCg2bbZO5v6fNAvbiXijjloJVu7RKDiEs3ItR'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [4]:
# Read Enhanced Twitter Archive Data
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [5]:
# Get Image Predictions file from Udacity's servers 
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

# Write contents to a file
with open('image-predictions.tsv', 'wb') as f:
    f.write(r.content)

In [6]:
# Read Image Predictions file
image_predictions = pd.read_csv('image-predictions.tsv', sep = '\t')

In [6]:
# Create a list of tweet IDs from Twitter Archive file
tweet_id_list = twitter_archive.tweet_id.values

In [7]:
# Run query on Twitter API to get tweet status for each tweet
start = timer()
with open('tweet_json.txt', 'w') as outfile:  
    for tweet_id in tweet_id_list:
        try:        
            tweet_status=api.get_status(tweet_id, tweet_mode='extended')            
            json.dump(tweet_status._json, outfile)
            outfile.write('\n')
        except:
            print("Tweet does not exist")
            pass
end = timer()
print(end - start)

Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Tweet does not exist
Rate limit reached. Sleeping for: 492
Tweet does not exist
Tweet does not exist
Rate limit reached. Sleeping for: 737
1667.420713163


In [7]:
# Read tweets JSON file into pandas data frame
tweet_status = pd.read_json('tweet_json.txt',lines=True)

**Note:** Alternate method

```python
with open('tweet_json.txt') as f:
    tweet_status = pd.DataFrame(json.loads(line) for line in f)
```

## Assess

In [48]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [51]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [52]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [57]:
twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [49]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [58]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [59]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [209]:
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Staffordshire_bullterrier     20
Siberian_husky                20
Cardigan                      19
web_site                      19
Eskimo_dog                    18
Maltese_dog                   18
teddy                         18
Shetland_sheepdog             18
beagle                        18
Rottweiler                    17
Lakeland_terrier              17
Shih-Tzu                      17
Italian_greyhound             16
kuvasz                        16
          

In [50]:
tweet_status

contributors  coordinates          created_at display_text_range  \
0              NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1              NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2              NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3              NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4              NaN          NaN 2017-07-29 16:00:24           [0, 138]   
5              NaN          NaN 2017-07-29 00:08:17           [0, 138]   
6              NaN          NaN 2017-07-28 16:27:12           [0, 140]   
7              NaN          NaN 2017-07-28 00:22:40           [0, 118]   
8              NaN          NaN 2017-07-27 16:25:51           [0, 122]   
9              NaN          NaN 2017-07-26 15:59:51           [0, 133]   
10             NaN          NaN 2017-07-26 00:31:25           [0, 130]   
11             NaN          NaN 2017-07-25 16:11:53           [0, 107]   
12             NaN          NaN 2017-07-25 01:55:32           [0, 106]   
13             NaN          NaN 2017-07-25 00:10:02            [0, 91]   
14             NaN          NaN 2017-07-24 17:02:04           [0, 118]   
15             NaN          NaN 2017-07-24 00:19:32           [0, 138]   
16             NaN          NaN 2017-07-23 00:22:39            [0, 86]   
17             NaN          NaN 2017-07-22 16:56:37           [0, 128]   
18             NaN          NaN 2017-07-22 00:23:06            [0, 87]   
19             NaN          NaN 2017-07-20 16:49:33           [0, 127]   
20             NaN          NaN 2017-07-19 16:06:48           [0, 127]   
21             NaN          NaN 2017-07-19 03:39:09           [0, 108]   
22             NaN          NaN 2017-07-19 00:47:34            [0, 99]   
23             NaN          NaN 2017-07-18 16:08:03            [0, 88]   
24             NaN          NaN 2017-07-18 00:07:08           [0, 129]   
25             NaN          NaN 2017-07-17 16:17:36           [0, 101]   
26             NaN          NaN 2017-07-16 23:58:41           [0, 121]   
27             NaN          NaN 2017-07-16 20:14:00            [0, 71]   
28             NaN          NaN 2017-07-15 23:25:31           [0, 131]   
29             NaN          NaN 2017-07-15 16:51:35          [27, 105]   
...            ...          ...                 ...                ...   
2307           NaN          NaN 2015-11-17 00:24:19           [0, 140]   
2308           NaN          NaN 2015-11-17 00:06:54           [0, 139]   
2309           NaN          NaN 2015-11-16 23:23:41           [0, 137]   
2310           NaN          NaN 2015-11-16 21:54:18            [0, 81]   
2311           NaN          NaN 2015-11-16 21:10:36           [0, 140]   
2312           NaN          NaN 2015-11-16 20:32:58           [0, 135]   
2313           NaN          NaN 2015-11-16 20:01:42           [0, 112]   
2314           NaN          NaN 2015-11-16 19:31:45           [0, 139]   
2315           NaN          NaN 2015-11-16 16:37:02           [0, 138]   
2316           NaN          NaN 2015-11-16 16:11:11           [0, 136]   
2317           NaN          NaN 2015-11-16 15:14:19            [0, 46]   
2318           NaN          NaN 2015-11-16 14:57:41            [0, 82]   
2319           NaN          NaN 2015-11-16 04:02:55           [0, 134]   
2320           NaN          NaN 2015-11-16 03:55:04           [0, 128]   
2321           NaN          NaN 2015-11-16 03:44:34           [0, 140]   
2322           NaN          NaN 2015-11-16 03:22:39           [0, 132]   
2323           NaN          NaN 2015-11-16 02:38:37           [0, 125]   
2324           NaN          NaN 2015-11-16 01:59:36           [0, 137]   
2325           NaN          NaN 2015-11-16 01:52:02           [0, 137]   
2326           NaN          NaN 2015-11-16 01:22:45           [0, 107]   
2327           NaN          NaN 2015-11-16 01:01:59           [0, 135]   
2328           NaN          NaN 2015-11-16 00:55:59           [0, 124]   
2329      

In [60]:
tweet_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2337 non-null datetime64[ns]
display_text_range               2337 non-null object
entities                         2337 non-null object
extended_entities                2065 non-null object
favorite_count                   2337 non-null int64
favorited                        2337 non-null bool
full_text                        2337 non-null object
geo                              0 non-null float64
id                               2337 non-null int64
id_str                           2337 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 n

#### Quality
##### `twitter_archive` table
- Nulls represented as 'None' in name, doggo, floofer, pupper amd puppo columns
- Some tweets are retweets 
- Some rating are dates such as 9/11 and 4/20
- Incorret dog names such as 'a', 'an', 'such', 'the','quite'.
- Tweets that are not ratings
- Erroneous datatypes (tweet_id, timestamp, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)


##### `image_predictions` table
- Erroneous datatypes (tweet_id)
- Predictions that are not dogs


##### `tweet_status` table
- Erroneous datatypes (id_str, in_reply_to_status_id_str, in_reply_to_user_id_str, quoted_status_id_str, favorite_count, retweet_count)

#### Tidiness
- Four columns (doggo, floofer, pupper, puppo) in `twitter_archive` table should be represented by one column (stage) 
- `image_predictions` table should be merged with `twitter_archive` table. There are 2356 tweets in `twitter_archive` table and 2075 image predictions in `image_predictions` . We only want tweets with images.
- favorite_count and retweet_count should be merged into `twitter_archive` table
- Drop unnecessary columns

## Cleaning

In [281]:
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean = image_predictions.copy()
tweet_status_clean = tweet_status.copy()

### Quality

#### Nulls represented as 'None' in name, doggo, floofer, pupper amd puppo columns

##### Define
Replace all None values with NaN

##### Code

In [282]:
# Covert the None values to Nan
twitter_archive_clean.name = twitter_archive_clean.name.replace('None', np.nan)
twitter_archive_clean.doggo = twitter_archive_clean.doggo.replace('None', np.nan)
twitter_archive_clean.floofer = twitter_archive_clean.floofer.replace('None', np.nan)
twitter_archive_clean.pupper = twitter_archive_clean.pupper.replace('None', np.nan)
twitter_archive_clean.puppo = twitter_archive_clean.puppo.replace('None', np.nan)

##### Test

In [283]:
twitter_archive_clean[twitter_archive_clean.name == 'None']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [284]:
twitter_archive_clean[twitter_archive_clean.doggo == 'None']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [285]:
twitter_archive_clean[twitter_archive_clean.floofer == 'None']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [286]:
twitter_archive_clean[twitter_archive_clean.pupper == 'None']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [287]:
twitter_archive_clean[twitter_archive_clean.puppo == 'None']

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

####  Some tweets are retweets 

##### Define
Only keep rows that are original tweets (i.e. drop retweets) then frop columns related to retweets (retweeted_status_id, retweeted_status_user_id, retweeted_status_id, retweeted_status_timestamp)

#### Code

In [288]:
# Remove rows that are retweets
twitter_archive_clean = twitter_archive_clean[pd.isna(twitter_archive_clean['retweeted_status_id'])]

#### Test

In [289]:
# Check the retweeted_status_id, retweeted_status_user_id & retweeted_status_timestamp are empty
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          1495 non-null object
doggo                         87 non-null object
floofer                       10 non-null object
pupper                        234 non-null object
puppo                         25 non-null object
dtypes: float64(4), int64(3), object(10)
memo

In [290]:
# Drop the retweet columns
twitter_archive_clean = twitter_archive_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_id', 'retweeted_status_timestamp'], axis=1)

In [291]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null object
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
rating_numerator         2175 non-null int64
rating_denominator       2175 non-null int64
name                     1495 non-null object
doggo                    87 non-null object
floofer                  10 non-null object
pupper                   234 non-null object
puppo                    25 non-null object
dtypes: float64(2), int64(3), object(9)
memory usage: 254.9+ KB


#### Some rating are dates such as 9/11 and 4/20

##### Define
In order to get accurate ratings, read through text column and only select ratings that have denominator of 10

#### Code

In [292]:
# Find the first rating in a tweet with a denominator of 10
def getRating(text,idx):
    regex = r'(?<=\s)([1-9][0-9]*\/[1-9][0-9]*)(?=\.|\s)'
    ratings = re.findall(regex, text)
    if ratings is not None:
        for rating in ratings:
            nums = rating.split('/')
            denom = nums[1]
            if denom == "10":
                return nums[idx]
    return np.nan

twitter_archive_clean.rating_numerator = twitter_archive_clean['text'].apply(getRating,idx=0)
twitter_archive_clean.rating_denominator = twitter_archive_clean['text'].apply(getRating,idx=1)

#### Test

In [293]:
twitter_archive_clean.rating_denominator.unique()

array(['10', nan], dtype=object)

In [294]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null object
source                   2175 non-null object
text                     2175 non-null object
expanded_urls            2117 non-null object
rating_numerator         2071 non-null object
rating_denominator       2071 non-null object
name                     1495 non-null object
doggo                    87 non-null object
floofer                  10 non-null object
pupper                   234 non-null object
puppo                    25 non-null object
dtypes: float64(2), int64(1), object(11)
memory usage: 254.9+ KB


#### Incorret dog names such as 'a', 'an', 'such', 'the','quite'

##### Define
Extract dog names from text column follwoing keywords such as 'name is', 'named', 'This is', 'Meet', 'Say hello to'

#### Code

In [295]:
# Find the dog's name in a tweet
def matchName(text):
    regex = r'(?<=(name is ))(([A-Z]){1}([a-z])+)(?=\.)|(?<=(named ))(([A-Z]){1}([a-z])+)(?=\.)|(?<=(This is ))(([A-Z]){1}([a-z])+)(?=\.)|(?<=(Say hello to ))(([A-Z]){1}([a-z])+)(?=\.)|(?<=(Meet ))(([A-Z]){1}([a-z])+)(?=\.)'
    result = re.search(regex,text)
    if result is not None:
        return result.group()
    return np.nan


twitter_archive_clean.name = twitter_archive_clean['text'].apply(matchName)

#### Test

In [296]:
twitter_archive_clean.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1139  728015554473250816                    NaN                  NaN   
187   856282028240666624                    NaN                  NaN   
1714  680440374763077632                    NaN                  NaN   
1072  739979191639244800                    NaN                  NaN   
237   847157206088847362                    NaN                  NaN   

                      timestamp  \
1139  2016-05-05 00:16:48 +0000   
187   2017-04-23 23:01:59 +0000   
1714  2015-12-25 17:30:01 +0000   
1072  2016-06-07 00:36:02 +0000   
237   2017-03-29 18:43:12 +0000   

                                                 source  \
1139  <a href="http://twitter.com/download/iphone" r...   
187   <a href="http://twitter.com/download/iphone" r...   
1714  <a href="http://twitter.com/download/iphone" r...   
1072  <a href="http://twitter.com/download/iphone" r...   
237   <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
1139  This is Rueben. He has reached ultimate pupper...   
187   This is Cermet, Paesh, and Morple. They are ab...   
1714  Merry Christmas. My gift to you is this tiny u...   
1072  This is Nollie. She's waving at you. If you do...   
237   Meet Daisy. She's been pup for adoption for mo...   

                                          expanded_urls rating_numerator  \
1139  https://twitter.com/dog_rates/status/728015554...               11   
187   https://twitter.com/dog_rates/status/856282028...               14   
1714  https://twitter.com/dog_rates/status/680440374...               11   
1072  https://twitter.com/dog_rates/status/739979191...               12   
237   https://www.petfinder.com/petdetail/37334596,h...               11   

     rating_denominator    name doggo floofer  pupper puppo  
1139                 10  Rueben   NaN     NaN  pupper   NaN  
187                  10     NaN   NaN     NaN     NaN   NaN  
1714                 10     NaN   NaN     NaN     NaN   NaN  
1072                 10  Nollie   NaN     NaN     NaN   NaN  
237                  10   Daisy   NaN     NaN     NaN   NaN

#### Tweets that are not ratings

##### Define
Drop tweets that are not ratings. Keep the rows that have a non-null denominator.

##### Code

In [297]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean['rating_denominator'].notnull()]

#### Test

In [298]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2071 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                 2071 non-null int64
in_reply_to_status_id    46 non-null float64
in_reply_to_user_id      46 non-null float64
timestamp                2071 non-null object
source                   2071 non-null object
text                     2071 non-null object
expanded_urls            2043 non-null object
rating_numerator         2071 non-null object
rating_denominator       2071 non-null object
name                     1303 non-null object
doggo                    84 non-null object
floofer                  10 non-null object
pupper                   231 non-null object
puppo                    24 non-null object
dtypes: float64(2), int64(1), object(11)
memory usage: 242.7+ KB


In [300]:
twitter_archive_clean.rating_denominator.nunique()

1

#### Erroneous datatypes (tweet_id, timestamp, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)

##### Define
Convert tweet_id, in_reply_to_status_id and in_reply_to_user_id to string data types. Convert  and timestamp to datetime data type.

#### Code

In [301]:
twitter_archive_clean.tweet_id = twitter_archive_clean.tweet_id.astype(str)
twitter_archive_clean.in_reply_to_status_id = twitter_archive_clean.in_reply_to_status_id.astype(str) 
twitter_archive_clean.in_reply_to_user_id  = twitter_archive_clean.in_reply_to_user_id.astype(str)
twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean['timestamp'])

#### Test

In [302]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2071 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                 2071 non-null object
in_reply_to_status_id    2071 non-null object
in_reply_to_user_id      2071 non-null object
timestamp                2071 non-null datetime64[ns]
source                   2071 non-null object
text                     2071 non-null object
expanded_urls            2043 non-null object
rating_numerator         2071 non-null object
rating_denominator       2071 non-null object
name                     1303 non-null object
doggo                    84 non-null object
floofer                  10 non-null object
pupper                   231 non-null object
puppo                    24 non-null object
dtypes: datetime64[ns](1), object(13)
memory usage: 242.7+ KB


#### Erroneous datatypes (tweet_id)

##### Define 
Convert tweet_id to string data type.

#### Code

In [303]:
image_predictions_clean.tweet_id = image_predictions_clean.tweet_id.astype(str)

#### Test

In [304]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


#### Drop predictions that are not dogs

##### Define
Only keep predictions that have number one predictions that are dogs.

#### Code

In [305]:
image_predictions_clean = image_predictions_clean[image_predictions_clean.p1_dog == True]

#### Test

In [306]:
image_predictions_clean[image_predictions_clean.p1_dog == False]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

#### Erroneous datatypes (id_str, in_reply_to_status_id_str, in_reply_to_user_id_str, quoted_status_id_str, favorite_count, retweet_count)

##### Define 
Convert id_str, in_reply_to_status_id_str, in_reply_to_user_id_str and quoted_status_id_str to string data type. 

Convert favorite_count and retweet_count to int data type. 

#### Code

In [307]:
tweet_status_clean.id_str = tweet_status_clean.id_str.astype(str)
tweet_status_clean.in_reply_to_status_id_str = tweet_status_clean.in_reply_to_status_id_str.astype(str)
tweet_status_clean.in_reply_to_user_id_str = tweet_status_clean.in_reply_to_user_id_str.astype(str)
tweet_status_clean.quoted_status_id_str = tweet_status_clean.quoted_status_id_str.astype(str)
tweet_status_clean.favorite_count = tweet_status_clean.favorite_count.astype(int)
tweet_status_clean.retweet_count = tweet_status_clean.retweet_count.astype(int)

#### Test

In [308]:
tweet_status_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2337 entries, 0 to 2336
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2337 non-null datetime64[ns]
display_text_range               2337 non-null object
entities                         2337 non-null object
extended_entities                2065 non-null object
favorite_count                   2337 non-null int64
favorited                        2337 non-null bool
full_text                        2337 non-null object
geo                              0 non-null float64
id                               2337 non-null int64
id_str                           2337 non-null object
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        2337 non-null object
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          23

### Tidiness

#### Four columns (doggo, floofer, pupper, puppo) in `twitter_archive` table should be one column (dog stage) 

##### Define
Combine the doggo, floofer, pupper and puppo columns into one column called stage. 

Change the new stage category to a category data type.

Drop the doggo, floofer, pupper and puppo columns.

##### Code

In [309]:
# Merge 4 dog stage columns into one
twitter_archive_clean['stage'] = twitter_archive_clean[['doggo', 'floofer', 'pupper', 'puppo']].fillna('').max(axis=1)
# Replace empty columns with NaN
twitter_archive_clean.stage = twitter_archive_clean.stage.replace('', np.nan)
# Convert stage column to type category
twitter_archive_clean.stage = twitter_archive_clean.stage.astype('category')
# Drop 4 stage columns(doggo, floofer, pupper, puppo)
twitter_archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1, inplace=True)

##### Test

In [310]:
# Check values in stage column
twitter_archive_clean.stage.value_counts()

pupper     231
doggo       72
puppo       24
floofer     10
Name: stage, dtype: int64

In [311]:
# Check columns were dropped
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2071 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id                 2071 non-null object
in_reply_to_status_id    2071 non-null object
in_reply_to_user_id      2071 non-null object
timestamp                2071 non-null datetime64[ns]
source                   2071 non-null object
text                     2071 non-null object
expanded_urls            2043 non-null object
rating_numerator         2071 non-null object
rating_denominator       2071 non-null object
name                     1303 non-null object
stage                    337 non-null category
dtypes: category(1), datetime64[ns](1), object(9)
memory usage: 180.2+ KB


#### `image_predictions` table should be merged with `twitter_archive` table. 
#### There are 2356 tweets in `twitter_archive` table and 2075 image predictions in `image_predictions` . We only want tweets with images.

##### Define
Merge twitter_archive table with image_predictions to only keep top dog prediction. 

Drop all other columns from image_predictions table. 

Rename p1_dog to breed.

##### Code

In [312]:
# Merge the tables
twitter_archive_images_clean = pd.merge(twitter_archive_clean, image_predictions_clean, on='tweet_id')
# Only keep top prediction dog breed
twitter_archive_images_clean.drop(['p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'], axis=1, inplace=True)
# Rename p1 column to breed
twitter_archive_images_clean.rename(columns={'p1': 'breed'}, inplace=True)

##### Test

In [313]:
twitter_archive_images_clean.sample(5)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
1137  674788554665512960                   nan                 nan   
1311  668989615043424256                   nan                 nan   
661   727644517743104000                   nan                 nan   
660   727685679342333952                   nan                 nan   
176   837820167694528512                   nan                 nan   

               timestamp                                             source  \
1137 2015-12-10 03:11:43  <a href="http://twitter.com/download/iphone" r...   
1311 2015-11-24 03:08:48  <a href="http://twitter.com/download/iphone" r...   
661  2016-05-03 23:42:26  <a href="http://twitter.com/download/iphone" r...   
660  2016-05-04 02:26:00  <a href="http://twitter.com/download/iphone" r...   
176  2017-03-04 00:21:08  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
1137  Say hello to Maggie. She's a Western Septic Do...   
1311  This is Bernie. He's taking his Halloween cost...   
661   Here's a doggo struggling to cope with the win...   
660   This is Cilantro. She's a Fellation Gadzooks. ...   
176   Here's a pupper before and after being asked "...   

                                          expanded_urls rating_numerator  \
1137  https://twitter.com/dog_rates/status/674788554...                9   
1311  https://twitter.com/dog_rates/status/668989615...                3   
661   https://twitter.com/dog_rates/status/727644517...               13   
660   https://twitter.com/dog_rates/status/727685679...               12   
176   https://twitter.com/dog_rates/status/837820167...               12   

     rating_denominator      name   stage  \
1137                 10    Maggie     NaN   
1311                 10    Bernie     NaN   
661                  10       NaN   doggo   
660                  10  Cilantro     NaN   
176                  10       NaN  pupper   

                                              jpg_url  img_num  \
1137  https://pbs.twimg.com/media/CV1Um8vWIAAmhQn.jpg        1   
1311  https://pbs.twimg.com/media/CUi6geuUYAIvE9n.jpg        1   
661   https://pbs.twimg.com/media/Chkc1BQUoAAa96R.jpg        2   
660   https://pbs.twimg.com/media/ChlCQg-VIAQ_8g4.jpg        1   
176   https://pbs.twimg.com/media/C6CI_jbVAAA3-a1.jpg        1   

                 breed  
1137  miniature_poodle  
1311               pug  
661     Great_Pyrenees  
660      Border_collie  
176   golden_retriever

In [314]:
twitter_archive_images_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1420 entries, 0 to 1419
Data columns (total 14 columns):
tweet_id                 1420 non-null object
in_reply_to_status_id    1420 non-null object
in_reply_to_user_id      1420 non-null object
timestamp                1420 non-null datetime64[ns]
source                   1420 non-null object
text                     1420 non-null object
expanded_urls            1420 non-null object
rating_numerator         1420 non-null object
rating_denominator       1420 non-null object
name                     968 non-null object
stage                    225 non-null category
jpg_url                  1420 non-null object
img_num                  1420 non-null int64
breed                    1420 non-null object
dtypes: category(1), datetime64[ns](1), int64(1), object(11)
memory usage: 156.9+ KB


#### favorite_count and retweet_count should be merged into `twitter_archive` table

##### Define
Merge only favorite_count and retweet_count columns to master table

##### Code

In [315]:
# Merge favorite and retweet columns into twitter_archive_images_clean table
twitter_archive_master_clean = pd.merge(twitter_archive_images_clean, tweet_status_clean[['favorite_count', 'retweet_count', 'id_str']], left_on='tweet_id', right_on='id_str', how='left')

# Drop id_str column
twitter_archive_master_clean.drop('id_str', axis=1, inplace=True)

##### Test

In [316]:
twitter_archive_master_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1420 entries, 0 to 1419
Data columns (total 16 columns):
tweet_id                 1420 non-null object
in_reply_to_status_id    1420 non-null object
in_reply_to_user_id      1420 non-null object
timestamp                1420 non-null datetime64[ns]
source                   1420 non-null object
text                     1420 non-null object
expanded_urls            1420 non-null object
rating_numerator         1420 non-null object
rating_denominator       1420 non-null object
name                     968 non-null object
stage                    225 non-null category
jpg_url                  1420 non-null object
img_num                  1420 non-null int64
breed                    1420 non-null object
favorite_count           917 non-null float64
retweet_count            917 non-null float64
dtypes: category(1), datetime64[ns](1), float64(2), int64(1), object(11)
memory usage: 179.1+ KB


In [317]:
twitter_archive_master_clean.sample(5)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
863   695816827381944320                   nan                 nan   
1292  669603084620980224                   nan                 nan   
1091  676617503762681856                   nan                 nan   
775   707021089608753152                   nan                 nan   
142   848690551926992896                   nan                 nan   

               timestamp                                             source  \
863  2016-02-06 03:50:33  <a href="http://twitter.com/download/iphone" r...   
1292 2015-11-25 19:46:30  <a href="http://twitter.com/download/iphone" r...   
1091 2015-12-15 04:19:18  <a href="http://twitter.com/download/iphone" r...   
775  2016-03-08 01:52:18  <a href="http://twitter.com/download/iphone" r...   
142  2017-04-03 00:16:10  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
863   Here's a dog enjoying a sunset. 11/10 would tr...   
1292  Very human-like. Cute overbite smile *finger t...   
1091  I promise this wasn't meant to be a cuteness o...   
775   This is Tucker. He's a Dasani Episcopalian. Go...   
142   Please stop sending in animals other than dogs...   

                                          expanded_urls rating_numerator  \
863   https://twitter.com/dog_rates/status/695816827...               11   
1292  https://twitter.com/dog_rates/status/669603084...               10   
1091  https://twitter.com/dog_rates/status/676617503...               13   
775   https://twitter.com/dog_rates/status/707021089...               12   
142   https://twitter.com/dog_rates/status/848690551...               12   

     rating_denominator    name   stage  \
863                  10     NaN     NaN   
1292                 10     NaN     NaN   
1091                 10     NaN  pupper   
775                  10  Tucker     NaN   
142                  10     NaN     NaN   

                                              jpg_url  img_num  \
863   https://pbs.twimg.com/media/CagJtjYW8AADoHu.jpg        1   
1292  https://pbs.twimg.com/media/CUroc7QW4AATIff.jpg        1   
1091  https://pbs.twimg.com/media/CWPUB9TWwAALPPx.jpg        1   
775   https://pbs.twimg.com/media/Cc_XtkRW8AEE7Fn.jpg        2   
142   https://pbs.twimg.com/media/C8cnjHuXsAAoZQf.jpg        1   

                      breed  favorite_count  retweet_count  
863              Pomeranian          3120.0         1226.0  
1292            Maltese_dog           953.0          367.0  
1091              Chihuahua          2963.0         1023.0  
775          cocker_spaniel          4212.0         1432.0  
142   flat-coated_retriever         25758.0         4481.0

#### Drop Unneeded columns 

##### Code

In [318]:
twitter_archive_master_clean.drop(['tweet_id', 'text', 'in_reply_to_status_id', 'in_reply_to_user_id', 'source', 'expanded_urls', 'jpg_url', 'img_num'], axis=1, inplace=True)

##### Test

In [319]:
twitter_archive_master_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1420 entries, 0 to 1419
Data columns (total 8 columns):
timestamp             1420 non-null datetime64[ns]
rating_numerator      1420 non-null object
rating_denominator    1420 non-null object
name                  968 non-null object
stage                 225 non-null category
breed                 1420 non-null object
favorite_count        917 non-null float64
retweet_count         917 non-null float64
dtypes: category(1), datetime64[ns](1), float64(2), object(4)
memory usage: 90.3+ KB


In [320]:
twitter_archive_master_clean.sample(5)

timestamp rating_numerator rating_denominator    name stage  \
1303 2015-11-25 03:18:15                8                 10     Dug   NaN   
277  2017-01-03 01:20:49               12                 10   Comet   NaN   
1290 2015-11-26 01:07:38               10                 10   Kloey   NaN   
446  2016-09-14 23:30:38               10                 10     NaN   NaN   
355  2016-11-09 16:22:22               12                 10  Snicku   NaN   

           breed  favorite_count  retweet_count  
1303   Chihuahua          2708.0         1289.0  
277   Pomeranian          9427.0         2328.0  
1290  Pomeranian           389.0          112.0  
446   Rottweiler             NaN            NaN  
355     Pekinese         11662.0         4513.0

### Storing, Analysis and Data Visialization

In [321]:
twitter_archive_master_clean.to_csv('twitter_archive_master.csv', encoding='utf-8', index=False)